nemo/collections/asr/modules/conformer_encoder.py

```diff
@@ -235,57 +321,116 @@ def set_max_audio_length(self, max_audio_length):
         max_audio_length: int = audio_signal.size(-1)
-
         if max_audio_length > self.max_audio_length:
             self.set_max_audio_length(max_audio_length)
-
         if length is None:
             length = audio_signal.new_full(
                 audio_signal.size(0), max_audio_length, dtype=torch.int32, device=self.seq_range.device
             )
 
+        if cache_last_channel is not None:
+            cache_last_time_next = torch.zeros_like(cache_last_time)
+        else:
+            cache_last_time_next = None
         audio_signal = torch.transpose(audio_signal, 1, 2)
 
         if isinstance(self.pre_encode, nn.Linear):
-            audio_signal = self.pre_encode(audio_signal)
+            audio_signal = self.pre_encode(x=audio_signal)
         else:
-            audio_signal, length = self.pre_encode(audio_signal, length)
+            audio_signal, length = self.pre_encode(x=audio_signal, lengths=length)
+            if self.streaming_cfg.drop_extra_pre_encoded > 0 and cache_last_channel is not None:
+                audio_signal = audio_signal[:, self.streaming_cfg.drop_extra_pre_encoded :, :]
+                # TODO: find a better solution
+                length = (length - self.streaming_cfg.drop_extra_pre_encoded).float()
+                length = torch.clip(length, min=0).int()
 
-        audio_signal, pos_emb = self.pos_enc(audio_signal)
-        # adjust size
         max_audio_length = audio_signal.size(1)
+
         # Create the self-attention and padding masks
+        if cache_last_channel is not None:
+            last_channel_num, bs, cache_len, channel_size = cache_last_channel.size()
+            cache_keep_size = max_audio_length - self.streaming_cfg.cache_drop_size
+            cache_last_channel_next = torch.zeros(
+                (last_channel_num, bs, cache_len + cache_keep_size, channel_size),
+                device=cache_last_channel.device,
+                dtype=cache_last_channel.dtype,
+            )
+            max_audio_length += cache_len
+            padding_length = length + cache_len
+            audio_signal, pos_emb = self.pos_enc(x=audio_signal, cache_len=cache_len)
+        else:
+            padding_length = length
+            cache_last_channel_next = None
+            audio_signal, pos_emb = self.pos_enc(x=audio_signal)
 
-        pad_mask = self.make_pad_mask(max_audio_length, length)
-        att_mask = pad_mask.unsqueeze(1).repeat([1, max_audio_length, 1])
-        att_mask = torch.logical_and(att_mask, att_mask.transpose(1, 2))
-        if self.att_context_size[0] >= 0:
-            att_mask = att_mask.triu(diagonal=-self.att_context_size[0])
-        if self.att_context_size[1] >= 0:
-            att_mask = att_mask.tril(diagonal=self.att_context_size[1])
-        att_mask = ~att_mask
+        pad_mask = self.make_pad_mask(max_audio_length=max_audio_length, seq_lens=padding_length)
```

Comment:
Add some comments regarding shapes or whats going on, I'm quite lost at this depth of the code.

Reply:
done.